### PR TITLE
Generate primary:C, subkey:S, subkey:EtEr on `sq-dsm key generate`

### PR DESCRIPTION
**Pull request by zugzwang**
- New structure on `sq-dsm key generate --dsm-key="..."` (see below)
- New directory of known keys, used for testing backwards compatibility, and script to populate it with current version.

Note:
Until now, sq-dsm has been generating keys with gpg-like structure:

- Primary key: Certification + Signing
- Subkey: Encryption (transport & rest)

```
# NOTE: Generated with gpg
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQGNBGPP8GgBDADkbOREAOeTp5GF27phminH/Co8afwsO5QSlt0XYz0wuIx4bt9a
evFng4f1yETxU6ijh02jv1VURKoT1cXMKVkYqwmbmrYbnMiv56MYgSVNDcNHIJUB
k6iyivdBy2iGiysb8+KdaDqhv90AMlSIPAoka9rqW9KC1b+cyz+BFhkHSQq3kHox
lXMcvAdCuQo/X7ZUbQsAz5K5GRwWAABY1x1LZ9Y/xOWV6CLOs9JlLB81gHNzZOkb
PCodVjVq0w/stno7/WymfpVF+GKsKNcbb6VHyDgxJHe1ASxD2XSAlskqZ3yUxn3m
8SSAG1/eRyRCgW3F82xtXZFuirBDx1Yn9j/XTUZ3iaAaz62A/NDMkYYXDq/euJDH
t6ox1ld8MEYbWTQm0fAF2OImSZ65SKS0E+CLSK1try7hYqrR6/dcVEEfzxbrLTEd
A9qFXoUxsTVTy5HxoM/eIEoIwxOdAAC/j5iCrf0kPzw+Dq2aRTrolXoj3P8unSp/
HgrDKdlLylGYon0AEQEAAbQmRnJhbmNpc2NvIChnZW5lcmF0ZWQgd2l0aCBncGcp
IDx4QHgueD6JAdQEEwEKAD4WIQTPuailds08vRxb2wGpqHhw0G07aQUCY8/waAIb
AwUJA8JnAAULCQgHAgYVCgkICwIEFgIDAQIeAQIXgAAKCRCpqHhw0G07aYqQC/48
F4cRlIexPnQfeTX87etIcB7xtgDQyYh6lI8xkD1EsbeLvWwsaaVC19ggh/+k6w7m
yYWAfOOqipj6VBC7n661eWzDxNaq6MSQiQ8P21T562uI5wuW4LjI4ffxg7H45f2/
62xo6Wq9yAvDZ29taW71zrkKFDn4ZU+QIR6FM9qCAbX4Dg0hAXH//tPutkjBFpl8
EJMgyg8jtYNbuzpJPjhyfODcQmxvVnqoF3MALB9GlPOA4BQc/vW1lCY1qDC6TH9B
dRMHny/KmjzDc3i5njCdP+e1CuTNzPNEjYuZAHhd94E4c/HRtywgek2LbSC3APs9
1g5ptERYvop5UmGX2sUfCvn5YLzoRbMeuk1PH9dX9sZuI+FPkdqnK+pO5X2Ae7ou
vf9IWw6vpEp2Ls9RKpOyuGfR3/gi1+8sAdUXXI8gJQH3m7errKFFD82vSwPcAXYM
BHJAU/M8QftPri3SQ5TvJxa/jgqlswboIYx+Lx3KQ3g3pgdFhp/Dun8tT2uTN4K5
AY0EY8/waAEMALGTZtASoieRI4Ut3yq0hYGvN9g2LL4bYHBnt1JGRzro0P5tvnLE
4L6QhKoKzrVglngwTyHNWajkbKjU7iKVa6QQCz8h3cS3mNwyGjM6yhYEdbaZfhc8
+iHpIq2MjTw94sOMJOtS+ILFY3LjRgvhJLCqjGLANbOahuqmVXWYO/C4EbTMSJ4o
hHlKwWFRbRWEKKis5ihrvKy5bS0WyRezirzCLGPV0ICJfWMZG/A5ACCt1fkF3dEN
kEr4T2HyPSsexqyN6QPWmzMpRhnBEoU2KvM0Gn1Fye/VV4XajA4ZUheu+AhJ6Uwj
75tbXYrlAhYLtSWHQ3u9lDnMnFbw2rFDyf4rxnN9nUhXZtLtBW4cOzQ0BTVjJIns
5RgKcAJ5/ZF6IXn9doGXBbIv/81LMm5AaK51vCgdUrXmxbLm6cDiMFO+7tP0lDWN
ooE/CSGQUxxwErWhsweGdvtX9RFVVgZuH9+VIB5d7POwNXJI9uRIULEU1xR4EIn0
jzVPpqKwLEQi9QARAQABiQG2BBgBCgAgFiEEz7mopXbNPL0cW9sBqah4cNBtO2kF
AmPP8GgCGwwACgkQqah4cNBtO2mNkAwAra+i2To8GkE1GQSAHpMIkRMdfrTeSpX5
3hBWc5+ip2sr8aG73EYHYEU+9ayoxbIL3PaLjcrwmlmjK3+2TA2H/nsraJAGKflr
/U2H1PfFZl0eO6cAiKpGq2Eid5naVH3R9fbpv1L6ojNJ4bHPlKcMZdHlPaTRFh3f
i6CyDCRJcSpVvnttXe8pM4Ct5WFZznee125gsPm39xoZxfLVjvnGMux6ORcuVuyQ
riPHffOE6JONKmWVraVWC3Clrbo3PHtcjW3+RTo84MmZS4EzaKWcwOorBsvYRUsZ
pstDVx3XZNw1HbQ2x41YfhWj1BVcCFaaOpo7pc5Z6r35fy3OsoaN3QxZPjckIveM
Jwwt9TSPQP2+ONRqXaQpKbq8Hb0OczJQD5rDiThltAjf7wkFIHDGJPOS+xFUicUl
EYCpn4C8OYEzhWnilLqQgQTotbZnWvD4iKXTRXAsR0FK+ybdXc4TkHJ4PZvJT/Za
Hk6aoEr60CtuSfrnwOoJ5OI4PiiamT0d
=TzAe
-----END PGP PUBLIC KEY BLOCK-----
```

This PR implements a Sequoia-like structure:

- Primary key: Certification
- Subkey 1: Signing
- Subkey 2: Encryption (transport & rest)

For instance,
```
## NOTE: Generated with `sq key generate`
-----BEGIN PGP PUBLIC KEY BLOCK-----
Comment: A2E7 45D0 5463 D5BE A275  89F0 1681 7CC3 0D60 8C6F
Comment: Francisco <x@x.x>

xjMEY9v8HhYJKwYBBAHaRw8BAQdA77wPg7Yf5aT+lLT1jHK47NxpN5mUNuVSGeHD
pKhOiXbCwBEEHxYKAIMFgmPb/B4FiQERsCUDCwkHCRAWgXzDDWCMb0cUAAAAAAAe
ACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5vcmeN8rRBGXu+Cg+nTeg6JcbR
/D4/vQC3rSZWKchiI9pYGAMVCggCmwECHgEWIQSi50XQVGPVvqJ1ifAWgXzDDWCM
bwAADhcA/01r6DENicgcs58fF49BmPl2/PyKcZNN8Cc+Bfmb6X9uAQCrXd6tT84o
r2SfR4onKAxGAUFSat2+M3c6j2CAYfoVAc0RRnJhbmNpc2NvIDx4QHgueD7CwBQE
ExYKAIYFgmPb/B4FiQERsCUDCwkHCRAWgXzDDWCMb0cUAAAAAAAeACBzYWx0QG5v
dGF0aW9ucy5zZXF1b2lhLXBncC5vcmcvHejy/tobFOtVU3SvmEIy40nP2tg3rz6U
HyfvxEP5PwMVCggCmQECmwECHgEWIQSi50XQVGPVvqJ1ifAWgXzDDWCMbwAAepEB
AKJVWbDyDZyEMUUFTniRji/nuMAVCuw0jW9PQn3DZa5XAP9+QlcfokdhlvB9C15Y
yCdDX0x7lo7HACY76vLlrJQ0A84zBGPb/B4WCSsGAQQB2kcPAQEHQMVrWI+zaqJR
lqOR6NVtm+GhM4PRqh5j2JFk3iUUkODswsDFBBgWCgE3BYJj2/weBYkBEbAlCRAW
gXzDDWCMb0cUAAAAAAAeACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5vcmfr
W/1P0Bnpn0j1twHd1mgC45omB3HNF5RpQMeNiRliYQKbAr6gBBkWCgBvBYJj2/we
CRBo+5ogfvMuv0cUAAAAAAAeACBzYWx0QG5vdGF0aW9ucy5zZXF1b2lhLXBncC5v
cme2lqAOMJHOtSTQim1KKOpAOIcc0xF7VZ0rBzLJozn15RYhBIwhFw+AjelX6UvI
gmj7miB+8y6/AAAEYQEA78xp/mJ8SBC6N0RKbgwbhguCsgT0uO2WqEvHWQGMUAAB
AKHbNbzEQ8zKjNUp8ndRpiWPx3ESlN+yUmI5WIlDolINFiEEoudF0FRj1b6idYnw
FoF8ww1gjG8AAJxiAQCukkbJU7Ezi/SgaseRC0dPSDTkqbOvgIfuxtjieGOi7gD/
Wwlp/bAEi7a5vB+rRFaw/WXktOI4lP6EMpEQI63qeA3OOARj2/weEgorBgEEAZdV
AQUBAQdA3KO89yE0r89EWkAlv3Zw8eQZXvdlBsaYnabt25kjmAEDAQgJwsAGBBgW
CgB4BYJj2/weBYkBEbAlCRAWgXzDDWCMb0cUAAAAAAAeACBzYWx0QG5vdGF0aW9u
cy5zZXF1b2lhLXBncC5vcmfqbDuw2Wb+tpVd4J5g/FXYU8C2NVsWK1cN6cDEjSx8
9AKbDBYhBKLnRdBUY9W+onWJ8BaBfMMNYIxvAAD6vAEAq8uBNCCz+SvxKTGz2033
qGVF/Vk0dcn32xs7nV1xbsoBAJ6oskzLDxfjurmXYkuV6PvVS+tkOm08gmmbVREv
x5wH
=k2Y7
-----END PGP PUBLIC KEY BLOCK-----
```

### CODE REVIEW

File: openpgp-dsm/src/lib.rs [hunk 1]
```diff
@@ -429,7 +429,22 @@ impl DsmAgent {
             }
         }
 
-        Err(anyhow::anyhow!(format!("Found no suitable signing key in DSM")))
+        Err(anyhow::anyhow!("Found no suitable signing key in DSM".to_string()))
+    }
+
+    fn new_signer_from_descriptor(credentials: Credentials, desc: &SobjectDescriptor) -> Result<Self> {
```
Comment (zugzwang) [inline openpgp-dsm/src/lib.rs:435]: Rename to `new_subkey_signer_from_descriptor`

File: sq/tests/dsm/generate_knownkeys_for_version.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+versionkeys="$SCRIPT_DIR/../data/knownkeys/sq-dsm-$v"
+
+# Generate keys with this version
+
+random=$(head /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w "10" | head -n 1)
+
+array=( rsa2k rsa3k rsa4k nistp256 nistp384 nistp521 cv25519 )
+for alg in "${array[@]}"
+do
+    dsm_name="generate-knownkeys-test-$random-$alg$"
```
Comment (xinyufort) [inline sq/tests/dsm/generate_knownkeys_for_version.sh:27]: Not that it matters, but is the extra $ at the end intentional?

File: sq/tests/dsm/generate_knownkeys_for_version.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    keyfile="$versionkeys-$alg.pgp"
+    user_id="Knownkey-Test-$alg (sq-dsm $v) <xyz@xyz.xyz>"
+    echo "Generating $keyfile..."
+    $sq key generate --userid="$user_id" --dsm-key="$dsm_name" --cipher-suite="$alg" --dsm-exportable
+    $sq key extract-dsm-secret --dsm-key="$dsm_name" > "$keyfile"
```
Comment (xinyufort) [inline sq/tests/dsm/generate_knownkeys_for_version.sh:32]: Is the goal to overwrite the knownkeys files, or to compare the knownkeys files with what is generated by sq-dsm?

**Review by xinyufort — APPROVED** [review]
approve, but do address my comments!

File: sq/tests/dsm/generate_knownkeys_for_version.sh (hunk 2 above)
Comment (zugzwang) [reply sq/tests/dsm/generate_knownkeys_for_version.sh:27]: Typo, does matter!

File: openpgp-dsm/src/lib.rs [hunk 4]
```diff
@@ -429,7 +429,24 @@ impl DsmAgent {
             }
         }
 
-        Err(anyhow::anyhow!(format!("Found no suitable signing key in DSM")))
+        Err(anyhow::anyhow!("Found no suitable signing key in DSM".to_string()))
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp-dsm/src/lib.rs:432]: Not necessary, you're right

File: openpgp-dsm/src/lib.rs [hunk 5]
```diff
@@ -811,7 +877,7 @@ pub fn generate_key(
             ..Default::default()
         };
         dsm_client.__update_sobject(
-            &subkey.uid()?, &update_req, "store PGP certificate as metadata"
+            &subkey.uid()?, &update_req, "store encryption subkey metadata"
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp-dsm/src/lib.rs:880]: +1

File: sq/tests/dsm/generate_knownkeys_for_version.sh (hunk 3 above)
Comment (zugzwang) [reply sq/tests/dsm/generate_knownkeys_for_version.sh:32]: The goal is to generate knownkeys (overwriting).
Then, later version CIs will use this keys for ensuring backwards compatibility.

File: openpgp-dsm/src/lib.rs [hunk 6]
```diff
@@ -374,7 +374,7 @@ impl DsmAgent {
         let descriptor = SobjectDescriptor::Name(key_name.to_string());
         let prim_sob = dsm_client
             .get_sobject(None, &descriptor)
-            .context(format!("could not get primary key {}", key_name))?;
+            .context(format!("could not get primary key {:?}", descriptor))?;
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp-dsm/src/lib.rs:377]: At some point we will switch to UUIDs instead of unstable key names, and this will be already refactored

**Comment by zugzwang** [comment]
Thanks for the review @xinyufort

**Pull request merged**